ddtrace/tracer/textmap.go

```diff
@@ -487,3 +501,292 @@ func (*propagatorB3) extractTextMap(reader TextMapReader) (ddtrace.SpanContext,
+			keyRgx.ReplaceAllString(k[len("_dd.p."):], "_"),
+			strings.ReplaceAll(valueRgx.ReplaceAllString(v, "_"), "=", "~"))
+		if b.Len()+len(tag) > 256 {
+			break
+		}
+		b.WriteString(";")
+		b.WriteString(tag)
+	}
+	// the old state is split by vendors, must be concatenated with a `,`
+	if len(oldState) == 0 {
+		return b.String()
+	}
+	for _, s := range strings.Split(strings.Trim(oldState, " \t"), ",") {
+		if strings.HasPrefix(s, "dd=") {
+			continue
+		}
+		listLength++
+		// if the resulting tracestateHeader exceeds 32 list-members,
+		// remove the rightmost list-member(s)
+		if listLength > 32 {
+			break
+		}
+		b.WriteString("," + strings.Trim(s, " \t"))
+	}
+	return b.String()
+}
+
+func (p *propagatorW3c) Extract(carrier interface{}) (ddtrace.SpanContext, error) {
+	switch c := carrier.(type) {
+	case TextMapReader:
+		return p.extractTextMap(c)
+	default:
+		return nil, ErrInvalidCarrier
+	}
+}
+
+func (*propagatorW3c) extractTextMap(reader TextMapReader) (ddtrace.SpanContext, error) {
+	var parentHeader string
+	var stateHeader string
+	// to avoid parsing tracestate header(s) if traceparent is invalid
+	if err := reader.ForeachKey(func(k, v string) error {
+		key := strings.ToLower(k)
+		switch key {
+		case traceparentHeader:
+			if parentHeader != "" {
+				return ErrSpanContextCorrupted
+			}
+			parentHeader = v
+		case tracestateHeader:
+			stateHeader = v
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	var ctx spanContext
+	if err := parseTraceparent(&ctx, parentHeader); err != nil {
+		return nil, err
+	}
+	if err := parseTracestate(&ctx, stateHeader); err != nil {
+		return nil, err
+	}
+	// ctx.traceID might be nil due to higher order bits, eg. 1....00000000
+	if ctx.spanID == 0 {
+		return nil, ErrSpanContextNotFound
+	}
+	return &ctx, nil
+}
+
+// parseTraceparent attempts to parse traceparentHeader which describes the position
+// of the incoming request in its trace graph in a portable, fixed-length format.
+// The format of the traceparentHeader is `-` separated string with in the
+// following format: `version-traceId-spanID-flags`,
+// where:
+// - version - represents the version of the W3C Tracecontext Propagation format in hex format.
+// - traceId - represents the propagated traceID in the format of 32 hex-encoded digits.
+// - spanID - represents the propagated spanID in the format of 16 hex-encoded digits.
+// - flags - represents the propagated flags in the format of 2 hex-encoded digits, and supports 8 unique flags.
+// Example value of HTTP `traceparent` header: `00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01`,
+// Currently, Go tracer doesn't support 128-bit traceIDs, so the full traceID (32 hex-encoded digits) must be
+// stored into a field that is accessible from the span’s context. TraceId will be parsed from the least significant 16
+// hex-encoded digits into a 64-bit number.
+func parseTraceparent(ctx *spanContext, header string) error {
+	nonWordRegex := "\\W_"
```

Comment:
I think this requires some explanation, since according to the format described in the function documentation, these characters shouldn't be in the header to begin with.

Reply:
They shouldn't be, true, but it is mostly aimed at corner cases. Also, some tests in system-tests/parametric test that the headers are only consisting of - and [a-f0-9]

Reply:
I'm a little confused, since this is called `...Regex`, but is used with the `string.Trim` function, which doesn't take a regex.

It looks to me like we're trimming the characters `\`, `W`, and `_`, which I think requires some explanation. How did you arrive at this character set? Why do we want to trim these, and only these characters?

Reply:
I finally saw 😅 thanks, fixing